examples/step-79/step-79.cc

```diff
@@ -0,0 +1,2756 @@
+    // of the box, respectively. The center region of the top
+    // boundary is given a separate boundary indicator: This is
+    // where we will apply the down force.
+    for (const auto &cell : triangulation.active_cell_iterators())
+      {
+        for (const auto &face : cell->face_iterators())
+          {
+            if (face->at_boundary())
+              {
+                const auto center = face->center();
+                if (std::fabs(center(1) - 0) < 1e-12)
+                  {
+                    face->set_boundary_id(BoundaryIds::no_force);
+                  }
+                else if (std::fabs(center(1) - 1) < 1e-12)
+                  {
+                    if ((std::fabs(center(0) - 3) < .3))
+                      {
+                        face->set_boundary_id(BoundaryIds::down_force);
+                      }
+                    else
+                      {
+                        face->set_boundary_id(BoundaryIds::no_force);
+                      }
+                  }
+                else if (std::fabs(center(0) - 0) < 1e-12)
+                  {
+                    face->set_boundary_id(BoundaryIds::no_force);
+                  }
+                else if (std::fabs(center(0) - 6) < 1e-12)
+                  {
+                    face->set_boundary_id(BoundaryIds::no_force);
+                  }
+              }
+          }
+      }
+  }
+
+
+  // Next, determine the constraints due to boundary values.  The
+  // bottom corners of the domain are kept in place in the $y$
+  // direction -- the bottom left also in the $x$ direction. deal.II
+  // generally thinks of boundary values as attached to pieces of the
+  // boundary, i.e., faces, rather than individual vertices. Indeed,
+  // mathematically speaking, one can not assign boundary values to
+  // individual points for the infinite-dimensional partial
+  // differential equation. But, since we are trying to reproduce a
+  // widely used benchmark, we will do so anyway and keep in mind that
+  // we have a finite-dimensional problem for which imposing boundary
+  // conditions at a single node is valid.
+  template <int dim>
+  void SANDTopOpt<dim>::setup_boundary_values()
+  {
+    boundary_values.clear();
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      {
+        for (const auto &face : cell->face_iterators())
+          {
+            if (face->at_boundary())
+              {
+                const auto center = face->center();
+
+                // Check whether the current face is on the bottom
+                // boundary, and if it is whether one of its
+                // vertices might be the bottom left or bottom
+                // right vertex:
+                if (std::fabs(center(1) - 0) < 1e-12)
+                  {
+                    for (unsigned int vertex_number = 0;
+                         vertex_number < GeometryInfo<dim>::vertices_per_cell;
```

Comment:
Let't not use `GeometryInfo`. here you could use `cell->vertex_indices()`

Reply:
I think this returns the global vertex index instead of the local one - unless you're saying there's a way to find the number of vertices a cell has with vertex_index()?

Reply:
I cannot get this to work, or find any documentation about cell->vertex_indices() or similar - I'm worried I'm looking in the wrong place

Reply:
https://www.dealii.org/developer/doxygen/deal.II/classTriaAccessor.html#ae40947cf863ecef1ee08ec93f411baaa

I guess you are not on the developer version of the documentation, are you?

Reply:
That makes sense. Thank you!